de.metas.business/src/main/java/de/metas/customs/CustomsInvoiceRepository.java

```diff
@@ -0,0 +1,229 @@
+		customsInvoiceRecord.setAD_User_ID(UserId.toRepoId(customsInvoice.getUserId()));
+
+		final DocTypeId docTypeId = customsInvoice.getDocTypeId();
+		customsInvoiceRecord.setC_DocType_ID(docTypeId.getRepoId());
+
+		customsInvoiceRecord.setDateInvoiced(TimeUtil.asTimestamp(customsInvoice.getInvoiceDate()));
+
+		customsInvoiceRecord.setDocAction(customsInvoice.getDocAction());
+		customsInvoiceRecord.setDocStatus(customsInvoice.getDocStatus().getCode());
+
+		customsInvoiceRecord.setC_Currency_ID(customsInvoice.getCurrencyId().getRepoId());
+
+		saveRecord(customsInvoiceRecord);
+
+		final CustomsInvoiceId customsInvoiceId = CustomsInvoiceId.ofRepoId(customsInvoiceRecord.getC_Customs_Invoice_ID());
+		customsInvoice.setId(customsInvoiceId);
+		customsInvoice.setCreatedBy(UserId.ofRepoId(customsInvoiceRecord.getCreatedBy()));
+		customsInvoice.setLastUpdatedBy(UserId.ofRepoId(customsInvoiceRecord.getUpdatedBy()));
+
+		for (final CustomsInvoiceLine line : customsInvoice.getLines())
+		{
+			final I_C_Customs_Invoice_Line existingLineRecord = existingLineRecords.remove(line.getId());
+			saveLine(line, customsInvoiceId, existingLineRecord);
+		}
+
+		//
+		// Delete remaining lines
+		deleteAll(existingLineRecords.values());
+
+		return customsInvoiceRecord;
+	}
+
+	private List<I_C_Customs_Invoice_Line> retrieveLineRecords(@NonNull final CustomsInvoiceId customsInvoiceId)
+	{
+		return Services.get(IQueryBL.class)
+				.createQueryBuilder(I_C_Customs_Invoice_Line.class)
+				.addEqualsFilter(I_C_Customs_Invoice_Line.COLUMN_C_Customs_Invoice_ID, customsInvoiceId)
+				.create()
+				.list();
+	}
+
+	private void saveLine(
+			@NonNull final CustomsInvoiceLine line,
+			@NonNull final CustomsInvoiceId customsInvoiceId,
+			final I_C_Customs_Invoice_Line existingRecord)
+	{
+		final I_C_Customs_Invoice_Line record;
+		if (existingRecord == null)
+		{
+			record = newInstance(I_C_Customs_Invoice_Line.class);
+		}
+		else
+		{
+			record = existingRecord;
+		}
+
+		record.setC_Customs_Invoice_ID(customsInvoiceId.getRepoId());
+
+		record.setLineNo(line.getLineNo());
+
+		final OrgId orgId = line.getOrgId();
+		record.setAD_Org_ID(orgId.getRepoId());
+
+		final Quantity quantity = line.getQuantity();
+		record.setInvoicedQty(quantity.getAsBigDecimal());
+		record.setC_UOM_ID(quantity.getUOMId());
+
+		final Money lineNetAmt = line.getLineNetAmt();
+		record.setLineNetAmt(lineNetAmt.getValue());
+
+		final ProductId productId = line.getProductId();
+		record.setM_Product_ID(productId.getRepoId());
+
+		record.setC_UOM_ID(line.getUomId().getRepoId());
+
+		saveRecord(record);
+
+		line.setId(CustomsInvoiceLineId.ofRepoId(record.getC_Customs_Invoice_ID(), record.getC_Customs_Invoice_Line_ID()));
+
+	}
+
+	public CustomsInvoice updateDocActionAndStatus(@NonNull final CustomsInvoice customsInvoice)
+	{
+		final I_C_Customs_Invoice customsInvoiceRecord = load(customsInvoice.getId(), I_C_Customs_Invoice.class);
```

Comment:
This occurs to me only now (a bit late) and i think we should go with it rather than doing another iteration, but:

As far as i understand DDD, in this case we should have one `CustomsInvoice` entity, all with lines and external references.
This repo should then provide one `save()` method and maybe one or more `getBy..(...)`. 
The other public methods like `setCustomsInvoiceToShipment`, `updateDocActionAndStatus` etc don't belong in there.
Responsibilities should be less distributed over `CustomsInvoiceRepository` and `CustomsInvoiceService`.

So, let's try to keep it in the back of our heads to next time, when writing something new or getting to touch this code again